src/Plugin/DataCollectorPlugin.php

```diff
@@ -166,14 +181,17 @@ public function detachFromMessageBus(MessageBus $messageBus): void
 
     protected function createContextFromActionEvent(ActionEvent $event): array
     {
+        /* @var $messageBus NamedMessageBus Is ensured above */
+        $messageBus = $event->getTarget();
+        $message = $event->getParam(MessageBus::EVENT_PARAM_MESSAGE);
+        $handler = $event->getParam(MessageBus::EVENT_PARAM_MESSAGE_HANDLER);
+
         return [
-            'bus-name' => $event->getTarget()->busName(),
-            'message-data' => $event->getParam('message')->toArray(),
-            'message-name' => $event->getParam('message-name'),
-            'message-handled' => $event->getParam('message-handled'),
-            'message-handler' => is_object($event->getParam('message-handler'))
-                ? get_class($event->getParam('message-handler'))
-                : $event->getParam('message-handler'),
+            'bus-name' => $messageBus->busName(),
+            'message-data' => $message instanceof DomainMessage ? $message->toArray() : [],
```

Comment:
You should use https://github.com/prooph/common/blob/master/src/Messaging/MessageConverter.php instead. Any class implementing Message interface should be good to go, not only those extending DomainMessage.

Reply:
Registering the `Prooph\Common\Messaging\NoOpMessageConverter` as default and add a (global?) configuration to exchange the message converter?

Reply:
sounds good to me

Reply:
maybe in a way, I can use different message converters for different event stores or contexts

Reply:
How about decorating the `NoOpMessageConverter` so that a non `DomainMessage` would result in an empty array instead of an exception (or `['please use a custom message converter']`)?
The conversion is just used for logging, therefore crashing just because you do net extend `DomainMessage` should not be necessary.

Reply:
Crashing is necessary because you chose the wrong message converter instance. You can try-catch in your plugin and display whatever you want to, but the NoOpMessageConverter should not change.

Reply:
> you can try-catch in your plugin and display whatever you want to

that's what I meant ;)

Reply:
the `Prooph\Common\Messaging\NoOpMessageConverter` requires the message to be an instance of `Prooph\Common\Messaging\Message` message which must not be the case.

Wrapping it with another converter?

Reply:
you can something like the following:
```
if (is_scalar($message) || is_array($message) { 
    $data = $message;
} elseif ($message instance of Message) {
    try {
        $data = $messageConverter->convertToArray($message); 
    } catch (RuntimeException $e) {
        $data = 'Cannot convert message to array, maybe you should use a custom MessageConverter?';
    }
} else {
    $data = sprintf(
        'can not extract message data, unknown message type %s given',
        is_object($message) ? get_class($message) : gettype($message)
    );
}
```

Reply:
It will take longer than expected to solve this accurately.

The conversion is necessary for two Plugins, the `PsrLoggerPlugin` and the `DataCollectorPlugin`.
But while there is one instance of the `PsrLoggerPlugin` for each message bus, there is only one instance of the `DataCollectorPlugin` for each message bus _type_.
This makes a consistent configuration harder.

Perhaps it could be an idea to split the `DataCollectorPlugin`, so that we still could have one `DataCollector` per message bus type, but one `Plugin` per message bus.

ping @lunetics

Reply:
or we just allow the user to disable those debugging plugins if they want to use custom messages. We cannot log or collect them anyway

Reply:
To get rid of this problem we would need to only allow `Prooph\Common\Messaging\DomainMessage` to be logged. Even `Prooph\Common\Messaging\Message` would require the possibility to exchange the `MessageConverter`.

Reply:
I think that's ok in this case. We are talking about debugging plugins and hey you can only use them if you use the prooph message implementation. Sure, would be nicer if we could provide an easy way to exchange MessageConverter and support the Message interface. We can put that in an issue for a future PR. At the moment we need to focus on getting the bundles stable so that they can be used in production.

So :+1: for supporting only `DomainMessage` at the moment but providing a way to disable debugging plugins if you want to use your own message implementation.

Reply:
Currently with the PR it should work fully on `DomainMessage` and partially on all other messages without throwing exceptions. I would let this as it is (+ fixing phpstan etc.) and do the rest in other PRs.

Reply:
agreed! We need to dig deeper into the `MessageConverter` issue. I'd say a global message converter used for all messages in the project is more than enough. But I do see value in the ability to configure a converter per bus type like suggested by @prolic